### PR TITLE
[BugFix] Fix memory leak in DGL edge softmax module

### PR DESCRIPTION
**Pull request by ylfdq1118**
## Description
Current DGL operators leak memory, and this happens for both PyTorch (#637 ) and MXNet backend.

For PyTorch, the cause is we used a separate cache instead of PyTorch's `ctx.save_for_backward` API to save tensors from forward pass but forgot to explicitly clear the cache. And the reason for not using `ctx.save_for_backward` API is we need to save non-tensor type data. Similar problem was mentioned in [this thread](https://discuss.pytorch.org/t/when-should-you-save-for-backward-vs-storing-in-ctx/6522).

For MXNet, the interesting thing is, even if we use its standard `save_for_backward` API, there is still memory leak. 

The Solution is simply explicitly clearing the cache.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

**Comment by SunderlandAJ-1130** [comment]
Hello, this problem still exists in DGL 0.6.1 with mxnet as backend. Could you please tell me how to fix it. Thanks!
